### ISSUE TITLE
[Bug] UWP CollectionView memory leak

### ISSUE THREAD

**Issue by JKennedy24** [xamarin/Xamarin.Forms#8928]
### Description
Navigating in and out of a page containing a collectionView on UWP keeps the collectionView and the whole page alive, causing a pretty big memory leak

### Steps to Reproduce

1.  Open sample on UWP
2.  Click "Mileage Page"
3.  Take snapshot
4. Click "Back"
5 Click "Mileage Page"
6. Click "Back"
7. Take SnapShot
8. Compare snapshot #2 to #1 and see `CollectionView` and all containing contents increases by 1. Repeat process N number of times to see CollectionView increase by N

![image](https://user-images.githubusercontent.com/9284438/70985449-064af800-20b4-11ea-8527-00d9909c3096.png)

### Expected Behavior
No memory leak

### Reproduction Link
[SfSample.zip](https://github.com/xamarin/Xamarin.Forms/files/3972596/SfSample.zip)

**Comment by JKennedy24** [comment]
My guess from my snapshotting is that an eventhandler is being left over.

The only conclusion I have is that the `TearDown` method for `CollectionView` isn't being called

https://github.com/xamarin/Xamarin.Forms/blob/88525dd913cf6ca54e162b0734a773c75454c385/Xamarin.Forms.Platform.UAP/CollectionView/ItemsViewRenderer.cs#L195

**Comment by JKennedy24** [comment]
OK so I overrode `CollectionViewRenderer` like so:

```
[assembly: ExportRenderer(typeof(CollectionView), typeof(FieldStrikeMove.UWP.CustomRenderers.Controls.MyCollectionViewRenderer))]
namespace FieldStrikeMove.UWP.CustomRenderers.Controls
{
    public class MyCollectionViewRenderer : CollectionViewRenderer
    {
        protected override void OnElementChanged(ElementChangedEventArgs<GroupableItemsView> args)
        {
            base.OnElementChanged(args);
        }

        protected override void TearDownOldElement(ItemsView oldElement)
        {
            var abc = Layout;   // Layout is null when page is popped. (therefore Layout.PropertyChanged is never unhooked)
            base.TearDownOldElement(oldElement);
        }
    }
```

And found that `Layout` was null during `TearDown`

This means that this event:
https://github.com/xamarin/Xamarin.Forms/blob/88525dd913cf6ca54e162b0734a773c75454c385/Xamarin.Forms.Platform.UAP/CollectionView/ItemsViewRenderer.cs#L205

Would never be unhooked.

I think that's what is keeping the control alive

**Comment by JKennedy24** [comment]
My workaround is to create this renderer

```
[assembly: ExportRenderer(typeof(CollectionView), typeof(FieldStrikeMove.UWP.CustomRenderers.Controls.ExtendedCollectionViewRenderer))]
namespace FieldStrikeMove.UWP.CustomRenderers.Controls
{
    public class ExtendedCollectionViewRenderer : CollectionViewRenderer
    {

        protected override void Dispose(bool disposing)
        {
            base.TearDownOldElement(Element);
            base.Dispose(disposing);
        }

    }
}
```

**Comment by hartez** [comment]
@JKennedy24 When I try to run your repro project, I'm getting `System.IO.FileNotFoundException: 'Could not load file or assembly 'libEGL, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.'`

**Comment by JKennedy24** [comment]
@hartez I don't get that error. I have never even heard of that library. Can you retry the sample again?

I don't think the sample matters much though. The collectionView has an obvious memory leak on UWP

**Comment by nschoenberg** [comment]
Can confirm the same is happening in a project im working on, using latest XF 4.4

**Comment by EmilAlipiev** [comment]
I think that this is related 

https://github.com/xamarin/Xamarin.Forms/issues/9051

**Comment by CliffAgius** [comment]
Anyone looking into this as it's a pretty big memory leak issue and we are seeing the same on a project using XF5.0.0.1709Pre4 making a Collection View unusable on UWP especially if it's containing an image.

**Comment by JKennedy24** [comment]
Heres another memory leak in the collectionview if anyone is interrested:

If you use an ObservableCollection to bind, You'll hit the following:

https://github.com/xamarin/Xamarin.Forms/blob/80a4dbfb1cd897f478c38d0aafa6a5d67940d2f3/Xamarin.Forms.Platform.UAP/CollectionView/GroupableItemsViewRenderer.cs#L30

https://github.com/xamarin/Xamarin.Forms/blob/80a4dbfb1cd897f478c38d0aafa6a5d67940d2f3/Xamarin.Forms.Platform.UAP/CollectionView/TemplatedItemSourceFactory.cs#L14

Cleanup never called in CollectionView
https://github.com/xamarin/Xamarin.Forms/blob/80a4dbfb1cd897f478c38d0aafa6a5d67940d2f3/Xamarin.Forms.Platform.UAP/CollectionView/ObservableItemTemplateCollection.cs#L53

**Comment by CliffAgius** [comment]
@JKennedy24 I have been looking at this over recent weeks but also trying a vanilla UPW project with an ObservableCollection so nothing Xamarin in the project the same Memory Leak is apparent.  There is various Forum Posts detailing this memory leak in the underlying UWP ListView, but I am trying to track it down and see if I can solve it.

However if you have a fix then please comment here or even better submit a PR. 😉

**Comment by JKennedy24** [comment]
If you want to workaround all memory leaks on UWP for `CollectionView` you must first implement the customrenderer mentioned here: https://github.com/xamarin/Xamarin.Forms/issues/8928#issuecomment-567014376 (you will need to add the ExtendedCollectionView as another ExportRenderer) 

and also use this extended version of CollectionView to get rid of memory leak mentioned here: https://github.com/xamarin/Xamarin.Forms/issues/8928#issuecomment-762378294

```
    public class ExtendedCollectionView : CollectionView, IDisposable
    {
        public void Dispose()
        {
            if (this.ItemsSource != null)
            {
                if (this.ItemsSource is INotifyCollectionChanged)
                {
                    var fi = this.ItemsSource.GetType().GetEventField("CollectionChanged");
                    if (fi == null) return;
                    fi.SetValue(this.ItemsSource, null);
                }
            }
        }
   }
```
Also using Extension Method:

```
        //https://stackoverflow.com/questions/43985099/how-to-remove-all-event-handler-from-an-observable-collections-collectionchange
        public static FieldInfo GetEventField(this Type type, string eventName)
        {
            FieldInfo field = null;
            while (type != null)
            {
                /* Find events defined as field */
                field = type.GetField(eventName, BindingFlags.Static | BindingFlags.Instance | BindingFlags.NonPublic);
                if (field != null && (field.FieldType == typeof(MulticastDelegate) || field.FieldType.IsSubclassOf(typeof(MulticastDelegate))))
                    break;

                /* Find events defined as property { add; remove; } */
                field = type.GetField("EVENT_" + eventName.ToUpper(), BindingFlags.Static | BindingFlags.Instance | BindingFlags.NonPublic);
                if (field != null)
                    break;
                type = type.BaseType;
            }
            return field;
        }
```

In my opinion alot of the "newer" controls in Xamarin.Forms are quite half baked (carouselView, collectionView etc..) and I avoid using them, opting for third party implementations instead (syncfusion, telerik, infragistics). Memory leaks seem to be completely overlooked/disregarded in these newer controls. Which unfortunately people who build real world applications cannot just disregard.

In this instance I was forced to use collectionview (unfortunately I dont have enough time to fix the source and submit PRs, If I did I would) so I am forced to work around the issues.

Hopefully this code helps someone else work around the memory issues in the collectionview in the uwp platform (other platforms not tested)

**Comment by YZahringer** [comment]
I think this issue is solved by #14780 

@JKennedy24 Could you check/validate it with the latest [5.0.0 Service Release 9](https://github.com/xamarin/Xamarin.Forms/releases/tag/release-5.0.0-sr9)?

**Comment by jfversluis** [comment]
No comment, going to assume this is fixed with #14780 thanks @YZahringer !
